cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/rolling/RackRolling.java

```diff
@@ -445,77 +249,26 @@ private static Set<KafkaNode> nextController(Reconciliation reconciliation,
-                            });
-                    kafkaNode.replicas().add(new Replica(
-                            replicatingBroker,
-                            topicDescription.name(),
-                            partition.partition(),
-                            partition.isr()));
-                });
-            });
-        });
+    private static Set<KafkaNode> nextNode(Reconciliation reconciliation, RollClient rollClient, Map<Integer, Context> contextMap, Map<Integer, NodeRoles> nodesNeedingRestart, int activeControllerId, int controllerCount, Map<Integer, Long> quorumState) {
+        KafkaNode brokerToRestart = null;
+        Map<Integer, KafkaNode> nodeIdToKafkaNode = Batching.nodeIdToKafkaNode(rollClient, contextMap, nodesNeedingRestart);
 
-        // Add any servers which we know about but which were absent from any partition metadata
-        // i.e. brokers without any assigned partitions
-        nodesNeedingRestart.forEach((nodeId, nodeRoles) -> nodeIdToKafkaNode.putIfAbsent(nodeId, new KafkaNode(nodeId, nodeRoles.controller(), nodeRoles.broker(), Set.of())));
-
-        // TODO somewhere in here we need to take account of partition reassignments
-        //      e.g. if a partition is being reassigned we expect its ISR to change
-        //      (see https://cwiki.apache.org/confluence/display/KAFKA/KIP-455%3A+Create+an+Administrative+API+for+Replica+Reassignment#KIP455:CreateanAdministrativeAPIforReplicaReassignment-Algorithm
-        //      which guarantees that addingReplicas are honoured before removingReplicas)
-        //      If there are any removingReplicas our availability calculation won't account for the fact
-        //      that the controller may shrink the ISR during the reassignment.
-
-        // Split the set of all brokers into subsets of brokers that can be rolled in parallel
-        var cells = cells(reconciliation, nodeIdToKafkaNode.values());
-        int cellNum = 0;
-        for (var cell: cells) {
-            LOGGER.debugCr(reconciliation, "Cell {}: {}", ++cellNum, cell);
-        }
+        var minIsrByTopic = rollClient.describeTopicMinIsrs(rollClient.listTopics().stream().map(TopicListing::name).toList());
 
-        // filter each cell by brokers that actually need to be restarted
-        cells = cells.stream()
-                .map(cell -> cell.stream()
-                        .filter(kafkaNode -> nodesNeedingRestart.containsKey(kafkaNode.id())).collect(Collectors.toSet()))
-                .toList();
-        cellNum = 0;
-        for (var cell: cells) {
-            LOGGER.debugCr(reconciliation, "Restart-eligible cell {}: {}", ++cellNum, cell);
+        for (KafkaNode kafkaNode : nodeIdToKafkaNode.values()) {
+            if (isQuorumHealthyWithoutNode(reconciliation, kafkaNode.id(), activeControllerId, controllerCount, quorumState)
+            && Batching.affectsAvailability(kafkaNode,minIsrByTopic)) {
```

Comment:
I hadn't paid that much attention to the naming of `affectsAvailability` before, but looks like naming is a bit confusing. If it does affect availability, we don't want to restart?